### PR TITLE
Add configurable gradient clipping norm

### PR DESCRIPTION
**Pull request by artest08**
## Summary

This PR makes the existing hard-coded training gradient clipping norm configurable through `clip_grad_norm`.

The default remains `10.0`, preserving current behavior, while `clip_grad_norm=0.0` disables gradient clipping. This keeps the change intentionally small.
## Changes

- Add `clip_grad_norm: 10.0` to `default.yaml`
- Register `clip_grad_norm` as a float config key
- Use `clip_grad_norm` in `BaseTrainer.optimizer_step()` instead of the hard-coded `10.0`
- Document the new training argument

## Motivation

In some of the Transformer decoder based architectures, this gradient clipping generally utilized as 0.1. Considering the future direction of the company for increasing the variations of the models, this might be a good step for paving way for it. 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds a new configurable `clip_grad_norm` training setting so users can control gradient clipping instead of relying on a fixed built-in value ⚙️

### 📊 Key Changes
- Added a new training argument, `clip_grad_norm`, to the documented training options 📝
- Added `clip_grad_norm: 10.0` to the default configuration in `default.yaml`
- Registered `clip_grad_norm` in config parsing so it is recognized as a valid numeric training parameter ✅
- Updated the trainer’s optimizer step to use the configured value instead of always clipping gradients at `10.0`
- Added support for disabling gradient clipping entirely by setting `clip_grad_norm` to `0.0` 🚫

### 🎯 Purpose & Impact
- Gives users more control over training stability, especially for models or datasets that are sensitive to large gradients 🎛️
- Makes training behavior easier to tune without editing source code, improving flexibility for research and production workflows
- Keeps the previous default behavior intact (`10.0`), so existing users should see no change unless they customize the setting 🔒
- Allows advanced users to disable clipping when needed, which can help with experimentation and optimizer tuning 🧪

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @artest08, thank you for submitting this `ultralytics/ultralytics` 🚀 PR for configurable gradient clipping! This is an automated message to help with review, and an engineer will assist you soon 😊 Please review the checklist below to help ensure a smooth integration:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -732,7 +732,9 @@ def setup_model(self):
     def optimizer_step(self):
         """Perform a single step of the training optimizer with gradient clipping and EMA update."""
         self.scaler.unscale_(self.optimizer)  # unscale gradients
-        torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=10.0)
+        clip_grad_norm = getattr(self.args, "clip_grad_norm", 10.0)
+        if clip_grad_norm and clip_grad_norm > 0:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:750]: 💡 **MEDIUM**: Negative values are silently treated the same as `0.0` here, even though the new option is documented as only using `0.0` to disable clipping. That makes a typo like `clip_grad_norm=-1` quietly turn clipping off and change training behavior. Please validate this argument as non-negative so invalid configs fail fast instead of being ignored.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change is small and well-scoped, and the default preserves existing behavior. One thing to tighten up before merge: reject negative `clip_grad_norm` values explicitly so misconfigurations don’t silently disable clipping.

💬 Posted 1 inline comment

**Comment by glenn-jocher** [comment]
Thanks! Your own follow-up #24420 (merged) documents the configurable gradient-clipping approach in the custom-trainer guide — a `clip_grad_norm` trainer-subclass override with disable-via-`0` semantics and a typical-values table (RT-DETR/DEIM/DETR `0.1`, YOLO `10.0`). That covers the need without adding a global config knob to the default args, which keeps the trainer surface minimal. Closing this in favor of that approach.
